### PR TITLE
Remodel swag requests to enable notifications

### LINKED ISSUE #363: Send swag requests via email
When a hackathon is approved AND it has requested swag, we need to send an email with the following:

- Organizer Name
- Hackathon Name
- Organizer Address

The email address receiving these requests is actually a Slack channel (`#hq-mail`). DM me for that email address.

After sending the email, the responsibility of getting the stickers to the organizer is off our hands. The Mail Team is responsible from there on.

---

We only want to send one swag envelope per approved hackathon. If a hackathon is approved multiple times, we don't want to send multiple envelopes. I'd suggest having some sort of way to track shipments. Possibly a `Swag::Shippment` model?? Here's some quick sudo code

```ruby
# requested_at datetime field
class Swag::Shippment < ApplicationModel
  after_create_commit :send_request

  belongs_to :hackathon
  has_one :mailing_address, through: :hackathon

  def send_request
    # send mailer
    touch :requested_at
  end

end
```

```ruby
# existing Hackathon model
has_one :swag_shippment

# when marking a hackathon as approved, create a `swag_shippment` if it doesn't already exist
```

We will not be building out a full mail/swag tracking system within Hackathons Backend. @recursiveforte is working on a mail tracking dashboard for Mail Team. Once their new mail tracking dashboard is built, we'll be sending a POST to request mail shipments (rather than emailing a Slack channel)

### PR DESCRIPTION
**Pull request by northeastprince**
Closes #363.

### CODE REVIEW

File: app/models/hackathon/swag_request/delivered.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+module Hackathon::SwagRequest::Delivered
+  extend ActiveSupport::Concern
+  extend Suppressible
+
+  def deliver_later_if_pertinent(wait: nil)
+    Hackathons::SwagRequestDeliveryJob.set(wait:).perform_later(self) if pertinent? && !suppressed?
+  end
+
+  def deliver_if_pertinent
+    HackathonMailer.with(hackathon:).swag_request.deliver if pertinent? && !suppressed?
```
Comment (garyhtou) [inline app/models/hackathon/swag_request/delivered.rb:10]: are we touching `delivered_at` anywhere?

File: app/views/hackathon_mailer/swag_request.text.erb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Hey mail team! 👋
+
+Hackathon swag has been requested for:
+
+<%= @hackathon.swag_request.mailing_address.to_s %>
```
Comment (garyhtou) [inline app/views/hackathon_mailer/swag_request.text.erb:5]: Please include the name of the Hackathon, and the applicant's name (if it is present). Ideally, we also format the address as multiline.

See https://github.com/hackclub/hackathons-backend/issues/363#issuecomment-1928512159

File: db/migrate/20240206140353_create_swag_requests.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+      t.references :mailing_address, null: false, foreign_key: true
+
+      t.timestamp :delivered_at
+
+      t.timestamps
+    end
+
+    Hackathon.where.associated(:swag_mailing_address).includes(:swag_mailing_address).find_each do |hackathon|
+      Hackathon::SwagRequest::Delivered.suppress do
+        hackathon.create_swag_request!(mailing_address: hackathon.swag_mailing_address.attributes)
```
Comment (garyhtou) [inline db/migrate/20240206140353_create_swag_requests.rb:14]: Would this create a brand new mailing address? Maybe we re-use the old one

Comment (northeastprince) [reply db/migrate/20240206140353_create_swag_requests.rb:14]: [We are.](https://github.com/hackclub/hackathons-backend/pull/426/files#:~:text=belongs_to%20%3Amailing_address)

File: app/models/hackathon/swag_request/delivered.rb (hunk 1 above)
Comment (northeastprince) [reply app/models/hackathon/swag_request/delivered.rb:10]: Ah, good catch! I'll touch it here.

File: app/views/hackathon_mailer/swag_request.text.erb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+Hey mail team! 👋
+
+Hackathon swag has been requested for:
+
+<%= @hackathon.applicant.name %> c/o <%= @hackathon.name %>
```
Comment (garyhtou) [inline app/views/hackathon_mailer/swag_request.text.erb:5]:
```suggestion
<%= @hackathon.name %><%= " c/o #{@hackathon.applicant.name}" if @hackathon.applicant.name %>
```

File: app/views/hackathon_mailer/swag_request.text.erb [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+<%= @mailing_address.line1 %>
+<%= @mailing_address.line2 %>
+<%= @mailing_address.city %><%= ", #{@mailing_address.province} #{@mailing_address.postal_code}" if @mailing_address.province %>
```
Comment (garyhtou) [inline app/views/hackathon_mailer/swag_request.text.erb:8]:
```suggestion
<% city, *province_and_postal = [@mailing_address.city, @mailing_address.province, @mailing_address.postal_code].compact_blank %>
<%= [city, province_and_postal.join(" ")].compact_blank.join(", ") %>
```

File: app/models/hackathon/swag_request/delivered.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+module Hackathon::SwagRequest::Delivered
+  extend ActiveSupport::Concern
+  extend Suppressible
+
+  def deliver_later_if_pertinent(wait: nil)
+    Hackathons::SwagRequestDeliveryJob.set(wait:).perform_later(self) if pertinent? && !suppressed?
```
Comment (garyhtou) [inline app/models/hackathon/swag_request/delivered.rb:6]: image

File: app/views/hackathon_mailer/swag_request.text.erb (hunk 4 above)
Comment (garyhtou) [reply app/views/hackathon_mailer/swag_request.text.erb:5]: We want the Hackathon's name first, then `c/o` the applicant's name

File: app/views/hackathon_mailer/swag_request.text.erb (hunk 5 above)
Comment (northeastprince) [reply app/views/hackathon_mailer/swag_request.text.erb:8]: What's the point of this?

Comment (garyhtou) [reply app/views/hackathon_mailer/swag_request.text.erb:8]: it handles the case where _any_ three of them are nil/blank. we can have addresses with a missing province or postal code!

**Review by garyhtou — APPROVED** [review]

**Comment by garyhtou** [comment]
I've just tested this integration in prod. It works!

image

**Pull request merged**